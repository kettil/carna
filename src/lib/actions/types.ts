import { Watermark, Watermarks } from 'istanbul-lib-report';
import { SpawnKillHandler } from '../utils/createSpawnKillHandler';

type BabelActionProps = {
  watch?: boolean;
  spawnKillHandler?: SpawnKillHandler;
};

type BabelNodeActionProps = {
  watch: boolean;
  script: string;
  watchPaths?: string[];
};

type CommitlintActionProps = {
  edit: string;
};

type CoverageWatermarkThreshold = Watermark | number | undefined;
type CoverageActionProps = {
  projects: string[];
  watermarks?: Partial<Record<keyof Watermarks, CoverageWatermarkThreshold>>;
};

type DepcheckActionProps = {
  path: string;
  ignorePackages?: string[];
};

type EslintActionProps = {
  write?: boolean;
  files?: string[];
};

type JestActionProps = {
  projects: string[];
  updateSnapshot?: boolean;
  runInBand?: boolean;
  verbose?: boolean;
  watch?: boolean;
};

type LicensecheckActionProps = {
  ignorePackages: string[];
  path?: string;
};

type PrettierActionProps = {
  write?: boolean;
  extension?: string;
  files?: string[];
};

type SemverActionProps = {
  path?: string;
};

type TscActionProps = {
  mode: 'type-check' | 'type-create';
};

export type {
  BabelActionProps,
  BabelNodeActionProps,
  CommitlintActionProps,
  CoverageActionProps,
  CoverageWatermarkThreshold,
  DepcheckActionProps,
  EslintActionProps,
  JestActionProps,
  LicensecheckActionProps,
  PrettierActionProps,
  SemverActionProps,
  TscActionProps,
};
