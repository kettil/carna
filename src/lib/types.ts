import { Logger } from './cli/logger';

export type NpmInstallMode = 'prod' | 'dev' | 'optional';

export type PropsGlobal = {
  readonly cwd: string;
  readonly tpl: string;
  readonly cfg: string;
  readonly log: Logger;
  readonly vvv: boolean;
};

export type Action<
  Props extends Record<string, unknown> | undefined = undefined,
  ActionReturn = void
> = Props extends undefined
  ? (argv: PropsGlobal) => Promise<ActionReturn>
  : (argv: PropsGlobal, props: Props) => Promise<ActionReturn>;
