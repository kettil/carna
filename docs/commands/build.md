# The `build` task

Build the application

## Options

| Options | Description       |
| ------- | ----------------- |
| --ci    | Run it in CI mode |

## Exit codes

| Exit code | Description |
| --------- | ----------- |
| 0         | Success     |
| 1         | task failed |
