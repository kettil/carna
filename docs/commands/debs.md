# The `debs` task

Checks if there are orphaned packages

## Options

| Options | Description       |
| ------- | ----------------- |
| --ci    | Run it in CI mode |

## Exit codes

| Exit code | Description |
| --------- | ----------- |
| 0         | Success     |
| 1         | task failed |
