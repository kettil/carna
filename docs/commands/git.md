# The `git` tasks

The tasks will be automatically called by husky during commit.

## The `git commit` task

Checks the files in staged with type, prettier and eslint.

### Exit codes

| Exit code | Description |
| --------- | ----------- |
| 0         | Success     |
| 1         | task failed |

## The `git msg` task

Checks the git commit message for a consistent structure.

### Options

| Options | Description                     |
| ------- | ------------------------------- |
| --edit  | Path to the COMMIT_EDITMSG file |

### Exit codes

| Exit code | Description |
| --------- | ----------- |
| 0         | Success     |
| 1         | task failed |

## The `git push` task

Checks the files with type, prettier and eslint.

### Exit codes

| Exit code | Description |
| --------- | ----------- |
| 0         | Success     |
| 1         | task failed |
