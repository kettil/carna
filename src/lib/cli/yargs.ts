import { CommandModule, Arguments, Argv } from 'yargs';
import ExecutableError from '../errors/executableError';
import { PropsGlobal } from '../types';

export type CommandModuleCommand = string;

export type CommandModuleDescribe = NonNullable<CommandModule<PropsGlobal>['describe']>;

export type CommandModuleBuilder<Props = Record<string, unknown>> = (
  yargs: Argv<PropsGlobal>,
) => Argv<PropsGlobal & Props>;

export type CommandModuleHandler<Props = Record<string, unknown>> = (args: Arguments<PropsGlobal & Props>) => void;

export const builderDefault = <Props>(cmd: string, fn: CommandModuleBuilder<Props>): CommandModuleBuilder<Props> => (
  yargs,
) => fn(yargs).usage(`Usage: $0 ${cmd} [options]`).help().version(false)
  .showHelpOnFail(false);

export const errorHandler = (argv: PropsGlobal, error: unknown): void => {
  if (error instanceof ExecutableError) {
    argv.log.log(error.stdout);

    /* eslint-disable-next-line node/no-process-exit, unicorn/no-process-exit */
    process.exit(1);
  }

  throw error;
};
