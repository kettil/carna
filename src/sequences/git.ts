import {
  CommandModuleBuilder,
  CommandModuleDescribe,
  CommandModuleCommand,
  CommandModuleHandler,
  builderDefault,
  errorHandler,
  commonHandler,
} from '../lib/cli/yargs';
import commit from './git/commit';
import msg, { Props as MessageProps } from './git/msg';
import push from './git/push';

export const command: CommandModuleCommand = 'git <hook>';
export const desc: CommandModuleDescribe = 'Handler for the git hooks';

const mode = ['msg', 'commit', 'push'] as const;
const options = { group: `${command.slice(0, Math.max(0, command.indexOf('<'))).trim()}-Options` } as const;

type Props = {
  readonly edit: MessageProps['edit'] | undefined;
  readonly hook: typeof mode[number];
};

export const builder: CommandModuleBuilder<Props> = builderDefault(command, (yargs) =>
  yargs.positional('hook', { choices: mode, demandOption: true }).options({
    edit: { ...options, type: 'string', desc: 'Path to the COMMIT_EDITMSG file (only at hook "msg")' },
  }),
);

export const handler: CommandModuleHandler<Props> = async (argv) => {
  try {
    const { edit, hook } = argv;

    await commonHandler(argv, false);

    switch (hook) {
      case 'commit':
        await commit(argv);
        break;

      case 'msg':
        if (typeof edit !== 'string' || edit.trim() === '') {
          throw new Error(`Argument "edit" is required at hook "${hook}"`);
        }

        await msg(argv, { edit });
        break;

      case 'push':
        await push(argv);
        break;
      default:
        throw new Error(`The hook "${hook}" is unknown`);
    }
  } catch (error) {
    errorHandler(argv, error);
  }
};
