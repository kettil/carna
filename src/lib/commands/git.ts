import { createBuilder, createHandler } from '../cli/yargs';
import type { GitProps } from '../tasks/gitTask';
import { hooks, gitTask } from '../tasks/gitTask';

const command = 'git';
const desc = 'Handler for the git hooks';

const options = { group: `${command}-Options` } as const;

const handler = createHandler<GitProps>(gitTask);
const builder = createBuilder<GitProps>(command, (yargs) =>
  yargs.options({
    hook: { ...options, choices: hooks, require: true },
    edit: {
      ...options,
      type: 'string',
      desc: 'Path to the COMMIT_EDITMSG file (only at hook "msg")',
    },
  }),
);

export { command, desc, builder, handler };
