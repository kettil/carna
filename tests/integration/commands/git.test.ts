import * as fs from 'fs';
import { Argv } from 'yargs';
import { handler, builder, command, desc } from '../../../src/lib/commands/git';
import { PropsGlobal } from '../../../src/lib/types';
import { getArgv } from '../../shared/setup/argv';
import { getReadFileWithHooksFiles } from '../../shared/setup/readFileFiles';

describe('command git', () => {
  test('it should be complete the yargs command structure', () => {
    expect(typeof command).toBe('string');
    expect(typeof desc).toBe('string');
    expect(typeof builder).toBe('function');
    expect(typeof handler).toBe('function');
  });

  test('it should be create the expected builder', () => {
    const yargs = {
      options: jest.fn((...args: unknown[]) => {
        expect(args).toMatchSnapshot('options');

        return yargs;
      }),
      usage: jest.fn((...args: unknown[]) => {
        expect(args).toMatchSnapshot('usage');

        return yargs;
      }),

      help: jest.fn(() => yargs),
      version: jest.fn(() => yargs),
      showHelpOnFail: jest.fn(() => yargs),
    } as unknown as Argv<PropsGlobal>;

    builder(yargs);
  });

  describe('with unknwon hook', () => {
    test('it should throw an error', async () => {
      const promise = handler(getArgv({ hook: 'foo' as any }));

      await expect(promise).rejects.toThrow('The git hook "foo" is unknown');
    });

    test('it should throw an error with hooks', async () => {
      (fs as any).setMockReadFileFiles(getReadFileWithHooksFiles());

      const promise = handler(getArgv({ hook: 'foo' as any, edit: 'path/to/commit.message' }));

      await expect(promise).rejects.toThrow('The git hook "foo" is unknown');
    });
  });
});
