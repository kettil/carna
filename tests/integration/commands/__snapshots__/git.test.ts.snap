// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`command git it should be create the expected builder: options 1`] = `
Array [
  Object {
    "edit": Object {
      "desc": "Path to the COMMIT_EDITMSG file (only at hook \\"msg\\")",
      "group": "git-Options",
      "type": "string",
    },
    "hook": Object {
      "choices": Array [
        "msg",
        "commit",
      ],
      "group": "git-Options",
      "require": true,
    },
  },
]
`;

exports[`command git it should be create the expected builder: usage 1`] = `
Array [
  "Usage: $0 git [options]",
]
`;

exports[`command git with unknwon hook it should throw an error with hooks: child-process 1`] = `
Object {
  "args": Array [
    "run",
    "precarna:git",
  ],
  "command": "npm",
  "cwd": "/path/to/project",
  "env": Object {},
}
`;

exports[`command git with unknwon hook it should throw an error with hooks: child-process 2`] = `
Object {
  "args": Array [
    "diff",
    "--staged",
    "--name-only",
  ],
  "command": "git",
  "cwd": "/path/to/project",
  "env": Object {},
}
`;

exports[`command git with unknwon hook it should throw an error with hooks: child-process 3`] = `
Object {
  "args": Array [
    "ls-files",
    "--exclude-standard",
    "--modified",
    "--others",
  ],
  "command": "git",
  "cwd": "/path/to/project",
  "env": Object {},
}
`;

exports[`command git with unknwon hook it should throw an error: child-process 1`] = `
Object {
  "args": Array [
    "diff",
    "--staged",
    "--name-only",
  ],
  "command": "git",
  "cwd": "/path/to/project",
  "env": Object {},
}
`;

exports[`command git with unknwon hook it should throw an error: child-process 2`] = `
Object {
  "args": Array [
    "ls-files",
    "--exclude-standard",
    "--modified",
    "--others",
  ],
  "command": "git",
  "cwd": "/path/to/project",
  "env": Object {},
}
`;
